src/util/routeUtils.js

```diff
@@ -1,74 +1,92 @@
+			m => (m.route.indexRoute ? [m.route.indexRoute] : [])
+		);
+	}
+	return resolveNestedRoutes(matchedRoute).then(m => m.route.routes || []);
+};
 
-/**
+const resolveNestedRoutes: Resolver<MatchedRoute> = matchedRoute =>
+	(matchedRoute.route.getNestedRoutes
+		? matchedRoute.route
+				.getNestedRoutes()
+				.then(routes => (matchedRoute.route.routes = routes))
+				.then(() => matchedRoute)
+		: Promise.resolve(matchedRoute));
+
+const resolveIndexRoute: Resolver<MatchedRoute> = matchedRoute =>
+	(matchedRoute.route.getIndexRoute
+		? matchedRoute.route
+				.getIndexRoute()
+				.then(indexRoute => (matchedRoute.route.indexRoute = indexRoute))
+				.then(() => matchedRoute)
+		: Promise.resolve(matchedRoute));
+
+/*
  * find all routes from a given array of route config objects that match the
- * supplied `url`
+ * supplied `location`
  *
- * this function matches the signature of `react-router-config`'s `matchRoutes`
+ * this function matches the signature of `react-router-config`'s `_resolveRouteMatches`
  * function, but interprets all `route.path` settings as nested
  *
  * @see {@link https://github.com/ReactTraining/react-router/tree/master/packages/react-router-config#matchroutesroutes-pathname}
- *
- * @param {Array} routes the routes to match
- * @param {String} url a URL path (no host) starting with `/`
- * @param {Array} matchedRoutes an array of [ route, match ] tuples
- * @param {String} matchedPath the part of the total path matched so far
- * @return {Array} an array of { route, match } objects
  */
-export const matchRoutes = (
-	routes = [],
-	url = '',
-	matchedRoutes = [],
-	matchedPath = ''
-) => {
-	const route = routes.find(r => matchPath(url, routePath(r, matchedPath))); // take the first match
-	if (!route) {
-		return matchedRoutes;
-	}
-
-	// add the route and its `match` object to the array of matched routes
-	const currentMatchedPath = routePath(route, matchedPath);
-	const match = matchPath(url, currentMatchedPath);
-	const currentMatchedRoutes = [...matchedRoutes, { route, match }];
-
-	// add any nested route matches
-	const nestedRoutes = getNestedRoutes({ route, match }) || [];
-	return matchRoutes(
-		nestedRoutes,
-		url,
-		currentMatchedRoutes,
-		currentMatchedPath
-	);
-};
-
-/**
- * @param {String} url the original request URL
- * @param {Array} queries an array of query function results
- * @param {Object} matchedRoute a { route, match } object to inspect for query functions
- * @return {Array} an array of returned query objects
+/*
+ * Collect the return values of each query functions associated with the matched
+ * routes, called with the provided location URL
+ *
+ * The function returned from calling `_matchedRouteQueriesReducer` with a
+ * `location` should be used as the callback to an `array.reduce` call.
  */
-export const matchedRouteQueriesReducer = location => (
-	queries,
-	{ route, match }
-) => {
+export const _matchedRouteQueriesReducer = (location: URL) => (
```

Comment:
ah, i need to drop the `export` because it is expected to _only_ be used internally by `getMatchedQueries`